pkg/dns/plugin.go

```diff
@@ -0,0 +1,284 @@
+				// Validation will ensure that the host is of the form *.foo.com
+				parts := strings.SplitN(host, ".", 2)
+				// Prefix wildcards with a . so that we can distinguish these entries in the map
+				key = fmt.Sprintf(".%s.", parts[1])
+			}
+			dnsEntries[key] = vips
+		}
+	}
+	h.mapMutex.Lock()
+	h.dnsEntries = make(map[string][]net.IP)
+	for k, v := range dnsEntries {
+		h.dnsEntries[k] = v
+	}
+	h.mapMutex.Unlock()
+}
+
+func convertToVIPs(addresses []string) []net.IP {
+	vips := make([]net.IP, 0)
+
+	for _, address := range addresses {
+		// check if its CIDR.  If so, reject the address unless its /32 CIDR
+		if strings.Contains(address, "/") {
+			if ip, network, err := net.ParseCIDR(address); err != nil {
+				ones, bits := network.Mask.Size()
+				if ones == bits {
+					// its a full mask (e.g., /32). Effectively an IP
+					vips = append(vips, ip)
+				}
+			}
+		} else {
+			if ip := net.ParseIP(address); ip != nil {
+				vips = append(vips, ip)
+			}
+		}
+	}
+
+	return vips
+}
+
+// ServerDNS is the implementation of DNS interface, when serving directly DNS-over-TLS
+func (h *IstioServiceEntries) ServeDNS(w dns.ResponseWriter, r *dns.Msg) {
+	response := h.handle(r)
+	var err error
+
+	if len(response.Answer) == 0 {
+		// Nothing found - forward to upstream DNS.
+		// TODO: this is NOT secured - pilot to k8s will need to use TLS or run a local coredns
+		// replica, using k8s plugin ( so this is over localhost )
+		response, _, err = h.client.Exchange(r, "")
+		if err != nil {
+			log.Debuga("DNS error ", r, err)
+			// TODO: increment counter - not clear if DNS client has metrics.
+		}
+	}
+
+	if len(response.Answer) == 0 {
+		response.Rcode = dns.RcodeNameError
+	}
+
+	err = w.WriteMsg(response)
+	if err != nil {
+		log.Debuga("DNS write error ", r, err)
+	}
+}
+
+func (h *IstioServiceEntries) handle(request *dns.Msg) *dns.Msg {
+	response := new(dns.Msg)
+	response.SetReply(request)
+	response.Authoritative = true
+
+	for _, q := range request.Question {
+		switch q.Qtype {
+		case dns.TypeA:
+			var vips []net.IP
+			h.mapMutex.RLock()
+			if h.dnsEntries != nil {
+				vips = h.dnsEntries[q.Name]
+				if vips == nil {
+					// check for wildcard format
+					// Split name into pieces by . (remember that DNS queries have dot in the end as well)
+					// Check for each smaller variant of the name, until we have
+					pieces := strings.Split(q.Name, ".")
+					pieces = pieces[1:]
+					for ; len(pieces) > 2; pieces = pieces[1:] {
+						if vips = h.dnsEntries[fmt.Sprintf(".%s", strings.Join(pieces, "."))]; vips != nil {
+							break
+						}
+					}
+				}
+			}
+			h.mapMutex.RUnlock()
+			if vips != nil {
+				response.Answer = a(q.Name, vips)
+			}
+		}
+	}
+	return response
+}
+
+// Query is used when acting as a coredns grpc plugin
+// code based on https://github.com/ahmetb/coredns-grpc-backend-sample
+func (h *IstioServiceEntries) Query(ctx context.Context, in *dnsapi.DnsPacket) (*dnsapi.DnsPacket, error) {
```

Comment:
We dont need this code because we are not going to act as a plugin. But if you want to make this act as a plugin, then you would also need a way to update the coredns config map with the cluster IP of Istiod as the target for *.global or some other DNS domain. Without that, there is no DNS forwarding. Unfortunately, if we do this procedure manually, it makes install step very brittle. We should do this in automated fashion, if we are making istiod a plugin.

Reply:
That will be in the next PR, if I can get this in :-)

There are few options - some are indeed brittle - changing kube-dns configmap is not that hard by itself, but will need a TLS cert and that's harder with citadel cert. 

The option I'm trying is to run coredns container next to istiod, with the k8s default config - 
in this mode:
- coredns will handle the DNS - we can compare perf/resources/etc with just using the small
DNS server in this PR, I don't want to make a decision without data.
- we don't have to worry about istiod -> kubedns - which is currently insecure.

As usual, 90% of the work will likely be around security - but I think we can get an 'out of box working' experience, as well as more advanced ones. 

Acting as a plugin will make sense in future, if we can get K8S to improve their DNS setup, 
adding a DNS-over-TLS by default. We still want to handle '.global' for example. So far the 
cost of having both is relatively small - if we find that it's not needed, easy to remove after we 
confirm the numbers.

Reply:
So let me make sure I understand the goal properly: the goal of this PR is to simply absorb the ecosystem coredns plugin into istiod. We would then run another coredns sidecar next to istiod. But this by itself will not do anything until the end user manually updates the kubedns config map  (old or new coredns format) to point to istiod service IP as an upstream DNS (for `*.global` or whatever). So essentially, in Kubernetes clusters where coredns is the default dns, the dns query path for something like .global will be

`pod --> foo.ns1.global --> k8s-core-dns ---> istiod-coredns-->istiod-coredns-plugin`

All this is in addition to the existing changes we do to the pod's DNS config to add .global as one of the DNS search domains in addition to .svc.cluster.local. 

Is this correct?